File: 0438-find-all-anagrams-in-a-string/0438-find-all-anagrams-in-a-string.py
class Solution:
    def findAnagrams(self, s: str, p: str) -> List[int]:

        ans = []

        if len(p) > len(s):
            return ans

        freqP = [0] * 26
        freqS = [0] * 26

        for ch in p:
            freqP[ord(ch) - ord('a')] += 1

        for i in range(len(p)):
            freqS[ord(s[i]) - ord('a')] += 1

        if freqP == freqS:
            ans.append(0)

        for i in range(len(p), len(s)):

            freqS[ord(s[i]) - ord('a')] += 1
            freqS[ord(s[i - len(p)]) - ord('a')] -= 1

            if freqP == freqS:
                ans.append(i - len(p) + 1)

        return ans

File: 0438-find-all-anagrams-in-a-string/0438-find-all-anagrams-in-a-string.cpp
class Solution {
public:
    vector<int> findAnagrams(string s, string p) {

        vector<int> ans;

        if (p.length() > s.length())
            return ans;

        vector<int> freqP(26, 0);
        vector<int> freqS(26, 0);

        for (char c : p)
            freqP[c - 'a']++;

        for (int i = 0; i < p.length(); i++)
            freqS[s[i] - 'a']++;

        if (freqP == freqS)
            ans.push_back(0);

        for (int i = p.length(); i < s.length(); i++) {

            freqS[s[i] - 'a']++;

            freqS[s[i - p.length()] - 'a']--;

            if (freqP == freqS)
                ans.push_back(i - p.length() + 1);
        }

        return ans;
    }
};